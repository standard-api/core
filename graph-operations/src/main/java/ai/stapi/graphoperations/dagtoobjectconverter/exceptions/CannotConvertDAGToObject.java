package ai.stapi.graphoperations.dagtoobjectconverter.exceptions;

import ai.stapi.graph.traversableGraphElements.TraversableNode;

public class CannotConvertDAGToObject extends RuntimeException {

    private CannotConvertDAGToObject(String message) {
        super("Cannot convert DAG to Object, because " + message);
    }
    
    public static CannotConvertDAGToObject becauseItContainsCycle(TraversableNode twiceVisitedNode) {
        return new CannotConvertDAGToObject(
            String.format(
                "it contains cycle. Twice visited node: '%s/%s'",
                twiceVisitedNode.getType(),
                twiceVisitedNode.getId()
            )
        );
    }
}
