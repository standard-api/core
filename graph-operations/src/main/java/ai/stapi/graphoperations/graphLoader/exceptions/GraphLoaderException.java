package ai.stapi.graphoperations.graphLoader.exceptions;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.utils.Stringifier;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import ai.stapi.identity.UniqueIdentifier;
import java.util.List;

public class GraphLoaderException extends RuntimeException {

  public GraphLoaderException(String becauseMessage) {
    super("Cannot load graph, " + becauseMessage);
  }

  private GraphLoaderException(String becauseMessage, Throwable cause) {
    super("Cannot load graph, " + becauseMessage, cause);
  }

  public static GraphLoaderException becauseThereIsNoGraphElementWithProvidedUuid(
      UniqueIdentifier uuid,
      GraphDescription graphDescription
  ) {
    return new GraphLoaderException(
        "because specified graph element does not exist in graph." +
            "\nUUID: " + uuid +
            "\nGraphDescription: " + graphDescription.getClass().getSimpleName() +
            "\n" + Stringifier.convertToString(graphDescription)
    );
  }

  public static GraphLoaderException becauseThereIsMoreThanOneNodeDescriptionsOnEdgeDescription(
      UniqueIdentifier id, String type) {
    return new GraphLoaderException(
        "because there is more than one node descriptions on edge description." +
            "\nUnique Identifier: " + id +
            "\nEdge type: " + type
    );
  }

  public static GraphLoaderException becauseEdgeDescriptionHadMoreChildNodeDescriptions(
      AbstractEdgeDescription edgeDescription,
      List<String> nodeTypes
  ) {
    var param = (EdgeDescriptionParameters) edgeDescription.getParameters();
    return new GraphLoaderException(
        "because edge description had more child node descriptions." +
            "\nEdge Description ClassName: " + edgeDescription.getClass().getSimpleName() +
            "\nEdge type: " + param.getEdgeType() +
            "\nNode Types: " + String.join(", ", nodeTypes)
    );
  }

  public static GraphLoaderException becauseProvidedOGMDidNotStartWithObjectOGM(
      ObjectGraphMapping objectGraphMapping) {
    return new GraphLoaderException(
        "because provided OGM did not start with object OGM." +
            "\nOGM Class: " + objectGraphMapping.getClass().getSimpleName()
    );
  }
}
