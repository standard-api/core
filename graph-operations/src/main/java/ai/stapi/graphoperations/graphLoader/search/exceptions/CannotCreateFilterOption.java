package ai.stapi.graphoperations.graphLoader.search.exceptions;

import ai.stapi.graphoperations.graphLoader.search.filterOption.FilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.FilterOptionParameters;

public class CannotCreateFilterOption extends RuntimeException {

  private CannotCreateFilterOption(String becauseMessage) {
    super("Cannot create FilterOption, " + becauseMessage);
  }

  public static CannotCreateFilterOption becauseInvalidLeafFilterStrategyProvided(
      String filterStrategy) {
    return new CannotCreateFilterOption(
        "because invalid leaf filter strategy provided. " +
            "Strategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidListFilterStrategyProvided(
      String filterStrategy) {
    return new CannotCreateFilterOption(
        "because invalid list filter strategy provided. " +
            "Strategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidLogicalFilterStrategyProvided(
      String filterStrategy) {
    return new CannotCreateFilterOption(
        "because invalid logical filter strategy provided. " +
            "Strategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidFilterOptionParameters(
      String filterStrategy,
      FilterOptionParameters filterOptionParameters
  ) {
    return new CannotCreateFilterOption(
        "because invalid filter option parameters provided. " +
            "\nParameters: " + filterOptionParameters.getClass().getSimpleName() +
            "\nStrategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidFilterOptionFoundWhenMakingCopy(
      FilterOption<?> originalFilter) {
    return new CannotCreateFilterOption(
        "because invalid filter option found when making copy. " +
            "\nFilter option: " + originalFilter.getClass().getSimpleName()
    );
  }
}
