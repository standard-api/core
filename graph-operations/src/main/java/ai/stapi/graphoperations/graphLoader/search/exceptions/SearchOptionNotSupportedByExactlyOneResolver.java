package ai.stapi.graphoperations.graphLoader.search.exceptions;

import ai.stapi.graphoperations.graphLoader.search.SearchOption;

public class SearchOptionNotSupportedByExactlyOneResolver extends RuntimeException {

  public SearchOptionNotSupportedByExactlyOneResolver(int size, SearchOption<?> searchOption) {
    super(
        String.format(
            "Search Option Of type: \"%s\" is supported by %d resolvers. But it should be exactly one.",
            searchOption.getClass().getCanonicalName(),
            size
        )
    );
  }
}
