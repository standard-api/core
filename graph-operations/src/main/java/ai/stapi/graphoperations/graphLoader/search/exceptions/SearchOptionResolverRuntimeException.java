package ai.stapi.graphoperations.graphLoader.search.exceptions;

public class SearchOptionResolverRuntimeException extends RuntimeException {

  private SearchOptionResolverRuntimeException(String becauseMessage) {
    super("Search option resolver threw exception, " + becauseMessage);
  }

  public static SearchOptionResolverRuntimeException becauseSearchQueryParamatersCanOnlyHaveOnePagingOptions() {
    return new SearchOptionResolverRuntimeException(
        "because Search Query Paramaters Can Only Have One Paging Options"
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideSortOptionMustBeSinglePath() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside Sort Option must be single path."
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideSortOptionMustEndWithValueDescription() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside Sort Option must end with value description."
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideOneValueFilterOptionMustBeSinglePath() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside OneValueFilterOption must be single path."
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideOneValueFilterOptionMustEndWithValueDescription() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside OneValueFilterOption must end with value description."
    );
  }

}
