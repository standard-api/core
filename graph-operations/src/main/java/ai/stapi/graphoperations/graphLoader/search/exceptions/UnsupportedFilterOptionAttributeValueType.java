package ai.stapi.graphoperations.graphLoader.search.exceptions;

public class UnsupportedFilterOptionAttributeValueType extends RuntimeException {

  public UnsupportedFilterOptionAttributeValueType(
      Class<?> attributeValueType,
      Class<?> optionClass
  ) {
    super(
        String.format(
            "Attribute value of type: \"%s\" cannot be used in: \"%s\"",
            attributeValueType.getCanonicalName(),
            optionClass.getCanonicalName()
        )
    );
  }

  public UnsupportedFilterOptionAttributeValueType(
      Class<?> optionClass
  ) {
    super(
        String.format(
            "Null attribute value cannot be used in: \"%s\"",
            optionClass.getCanonicalName()
        )
    );
  }
}
