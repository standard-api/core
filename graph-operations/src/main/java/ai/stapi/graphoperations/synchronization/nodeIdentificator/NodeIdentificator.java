package ai.stapi.graphoperations.synchronization.nodeIdentificator;

import ai.stapi.graphoperations.graphLanguage.graphDescription.graphDescriptionBuilder.GraphDescriptionBuilder;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractNodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.AttributeQueryDescription;
import ai.stapi.graphoperations.synchronization.nodeIdentificator.exception.InvalidNodeIdentificator;
import java.io.Serializable;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class NodeIdentificator implements Serializable {
    
    private final List<PositiveGraphDescription> graphDescriptions;
    
    public NodeIdentificator(String... attributeNames) {
        this(Arrays.stream(attributeNames).map(AttributeQueryDescription::new).collect(Collectors.toList()));
    }

    public NodeIdentificator(PositiveGraphDescription graphDescription) {
        this(List.of(graphDescription));
    }

    public NodeIdentificator(List<PositiveGraphDescription> graphDescriptions) {
        var invalidStart = graphDescriptions.stream()
            .filter(AbstractNodeDescription.class::isInstance)
            .toList();
        
        if (!invalidStart.isEmpty()) {
            throw InvalidNodeIdentificator.becausePathToIdentifiactionValueDoesntStartWithEdgeOrAttributeDescription(
                invalidStart
            );
        }
        var invalidBranching = graphDescriptions.stream()
            .filter(description -> !GraphDescriptionBuilder.isGraphDescriptionSinglePath(description))
            .toList();

        if (!invalidBranching.isEmpty()) {
            throw InvalidNodeIdentificator.becausePathToIdentifiactionValueIsNotSinglePath(
                invalidBranching
            );
        }
        var invalidEnd = graphDescriptions.stream()
            .filter(desc -> !GraphDescriptionBuilder.isGraphDescriptionEndingWithAttributeOrUuidDescription(desc))
            .toList();
        
        if (!invalidEnd.isEmpty()) {
            throw InvalidNodeIdentificator.becausePathToIdentifiactionValueIsNotEndingWithUuidIdentityOrAttributeDescription(
                invalidEnd
            );
        }
        this.graphDescriptions = graphDescriptions;
    }

    public List<PositiveGraphDescription> getGraphDescriptions() {
        return graphDescriptions;
    }
}
