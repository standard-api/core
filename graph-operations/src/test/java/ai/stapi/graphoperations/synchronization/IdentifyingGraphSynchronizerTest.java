package ai.stapi.graphoperations.synchronization;

import static org.junit.jupiter.api.Assertions.assertThrows;

import ai.stapi.graph.AttributeContainer;
import ai.stapi.graph.EdgeRepository;
import ai.stapi.graph.Graph;
import ai.stapi.graph.NodeRepository;
import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graph.exceptions.EdgeNotFound;
import ai.stapi.graph.exceptions.GraphException;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graph.repositorypruner.RepositoryPruner;
import ai.stapi.graphoperations.fixtures.testsystem.TestSystemModelDefinitionsLoader;
import ai.stapi.graphoperations.synchronization.nodeIdentificator.NodeIdentificator;
import ai.stapi.graphoperations.synchronization.nodeIdentificator.testImplementation.TestNodeIdentificatorsProvider;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope({
    IdentifyingGraphSynchronizerTestStructureLoader.SCOPE,
    TestSystemModelDefinitionsLoader.SCOPE
})
class IdentifyingGraphSynchronizerTest extends SchemaIntegrationTestCase {

    @Autowired
    protected RepositoryPruner repositoryPruner;
    
    @Autowired
    private NodeRepository nodeRepository;

    @Autowired
    private EdgeRepository edgeRepository;

    @Autowired
    private IdentifyingGraphSynchronizer identifyingGraphSynchronizer;
    
    @Autowired
    private TestNodeIdentificatorsProvider testNodeIdentificatorsProvider;

    @BeforeEach
    public void setUp() {
        this.repositoryPruner.prune();
    }

    @Test
    void itShouldSynchronizeNodeToEmptyGraph() throws GraphException {
        var expectedNode = new Node("test_node_type");

        var graph = new Graph(expectedNode);

        this.identifyingGraphSynchronizer.synchronize(graph);

        var actualNode = this.nodeRepository.loadNode(
            expectedNode.getId(),
            expectedNode.getType()
        );

        this.thenNodeApproved(actualNode);
    }

    @Test
    void itShouldSynchronizeNodeById() throws GraphException {
        var existingNode = new Node("Test_node_type");
        this.nodeRepository.save(existingNode);

        var expectedNode = new Node(
            existingNode.getId(),
            "Test_node_type"
        );
        expectedNode = expectedNode.add(
            new LeafAttribute<>("string_type", new StringAttributeValue("test value"
            ))
        );

        this.identifyingGraphSynchronizer.synchronize(new Graph(expectedNode));

        var actualNode = this.nodeRepository.loadNode(
            expectedNode.getId(),
            expectedNode.getType()
        );

        this.thenNodeApproved(actualNode);
    }

    @Test
    void itShouldNotSynchronizeNodeByIdAndDifferentType() throws GraphException {
        var existingNode = new Node("OriginalType");
        this.nodeRepository.save(existingNode);

        var expectedNode = new Node(
            existingNode.getId(),
            "DifferentType"
        );
        expectedNode = expectedNode.add(
            new LeafAttribute<>(
                "string_type", 
                new StringAttributeValue("test value")
            )
        );

        this.identifyingGraphSynchronizer.synchronize(new Graph(expectedNode));

        var actualNode = this.nodeRepository.loadNode(
            expectedNode.getId(),
            expectedNode.getType()
        );
        
        Assertions.assertTrue(this.nodeRepository.nodeExists(existingNode.getId(), existingNode.getType()));
        this.thenNodeApproved(actualNode);
    }


    @Test
    void itShouldNotFixEdgeWhenTryingToSynchronizeNodeWithSameAttributeButDifferentType() throws GraphException {
        var existingNode = new Node("SameType");
        var otherNode = new Node("OtherType");
        otherNode = otherNode.add(
            new LeafAttribute<>(
                "identifying_attribute",
                new StringAttributeValue("matching_value")
            )
        ).add(
            new LeafAttribute<>(
                "original_attribute",
                new StringAttributeValue("original_value")
            )
        );
        var anotherNode = new Node("AnotherType");
        var existingEdge = new Edge(existingNode, "originalEdge", otherNode);
        this.nodeRepository.save(existingNode);
        this.nodeRepository.save(otherNode);
        this.nodeRepository.save(anotherNode);
        this.edgeRepository.save(existingEdge);
        
        var mergingNode1 = new Node(
            existingNode.getId(),
            "SameType"
        );
        mergingNode1 = mergingNode1.add(
            new LeafAttribute<>(
                "string_type",
                new StringAttributeValue("test value")
            )
        );

        this.testNodeIdentificatorsProvider.add(
            otherNode.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        var mergingNode2 = new Node(anotherNode.getId(), "OtherType");
        mergingNode2 = mergingNode2.add(
            new LeafAttribute<>(
                "identifying_attribute",
                new StringAttributeValue("matching_value")
            )
        ).add(
            new LeafAttribute<>(
                "new_attribute",
                new StringAttributeValue("irrelevant_value")
            )
        );
        var newEdge = new Edge(mergingNode1, "originalEdge", mergingNode2);
        var someEdge = new Edge(mergingNode2, "originalEdge", anotherNode);
        var graph = new Graph(mergingNode1, mergingNode2, anotherNode, newEdge, someEdge);

        this.identifyingGraphSynchronizer.synchronize(graph);
        Assertions.assertTrue(this.nodeRepository.nodeExists(existingNode.getId(), existingNode.getType()));
        Assertions.assertTrue(this.nodeRepository.nodeExists(otherNode.getId(), otherNode.getType()));
        Assertions.assertTrue(this.nodeRepository.nodeExists(anotherNode.getId(), anotherNode.getType()));
        Assertions.assertTrue(this.edgeRepository.edgeExists(existingEdge.getId(), existingEdge.getType()));
        Assertions.assertFalse(this.nodeRepository.nodeExists(mergingNode2.getId(), mergingNode2.getType()));
    }

    @Test
    void itShouldSynchronizeNodeByIdAndMergeAttributes() throws GraphException {
        var existingNode = new Node("Test_node_type");
        var existingAttribute = new LeafAttribute<>("string_type", new StringAttributeValue("test value"
        ));
        existingNode = existingNode.add(
            existingAttribute
        );
        this.nodeRepository.save(existingNode);

        var synchronizedNode = new Node(
            existingNode.getId(),
            "Test_node_type"
        );
        var synchronizedAttribute =
            new LeafAttribute<>("string_type2", new StringAttributeValue("test value2"
            ));
        synchronizedNode = synchronizedNode.add(
            synchronizedAttribute
        );

        this.identifyingGraphSynchronizer.synchronize(new Graph(synchronizedNode));

        var actualNode = this.nodeRepository.loadNode(
            synchronizedNode.getId(),
            synchronizedNode.getType()
        );
        this.thenNodeApproved(actualNode);
    }

    @Test
    void itShouldSynchronizeNodeByIdAndMergeAttributesWithOverrideExisting(
    ) throws GraphException {
        var existingNode = new Node("Test_node_type");
        existingNode = existingNode.add(
            new LeafAttribute<>(
                "string_type",
                new StringAttributeValue("test value"))
        );
        this.nodeRepository.save(existingNode);

        var updatingNode = new Node(
            existingNode.getId(),
            "Test_node_type"
        );
        updatingNode = updatingNode.add(
            new LeafAttribute<>(
                "string_type",
                new StringAttributeValue("test value_overwritten"))
        );
        updatingNode = updatingNode.add(
            new LeafAttribute<>(
                "string_type2",
                new StringAttributeValue("test value2"))
        );

        this.identifyingGraphSynchronizer.synchronize(new Graph(updatingNode));

        var actualNode = this.nodeRepository.loadNode(
            existingNode.getId(),
            existingNode.getType()
        );

        this.thenNodeApproved(actualNode);
    }

    @Test
    void itShouldSynchronizeNodeByAttribute() throws GraphException {
        var updatingNodeType = "Test_node_type";
        var existingNode = new Node(updatingNodeType);
        existingNode = existingNode.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("identified value"
            ))
        );
        this.nodeRepository.save(existingNode);

        var updatingNode = new Node(updatingNodeType);
        
        this.testNodeIdentificatorsProvider.add(
            updatingNodeType, 
            new NodeIdentificator("identifying_attribute")
        );
        updatingNode = updatingNode.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("identified value"
            ))
        );
        updatingNode = updatingNode.add(
            new LeafAttribute<>("some_other_attribute",
                new StringAttributeValue("new synchronized value"
                ))
        );
        var graph = new Graph(updatingNode);
        this.identifyingGraphSynchronizer.synchronize(graph);
        var actualNode = this.nodeRepository.loadNode(
            existingNode.getId(),
            existingNode.getType()
        );
        this.thenNodeApproved(actualNode);
    }

    @Test
    void itShouldNotSynchronizeNodeByDifferentAttribute() throws GraphException {
        var notMergedNodeType = "Test_node_type";
        var existingNode = new Node(notMergedNodeType);
        existingNode = existingNode.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("identified value"
            ))
        );
        this.nodeRepository.save(existingNode);

        var notMergedNode = new Node(notMergedNodeType);
        this.testNodeIdentificatorsProvider.add(
            notMergedNodeType,
            new NodeIdentificator("identifying_attribute")
        );
        notMergedNode = notMergedNode.add(
            new LeafAttribute<>("not_identifying_attribute", new StringAttributeValue("test value"
            ))
        );

        this.identifyingGraphSynchronizer.synchronize(new Graph(notMergedNode));

        var actualNode = this.nodeRepository.loadNode(
            notMergedNode.getId(),
            notMergedNode.getType()
        );
        this.thenNodeApproved(actualNode);
    }

    @Test
    void itShouldSynchronizeEdgeWithItsNodesToEmptyGraph() throws GraphException {
        var nodeA = new Node("glass");
        var nodeB = new Node("water");

        var expectedEdge = new Edge(
            nodeA,
            "contains",
            nodeB
        );

        this.whenSynchronize(
            nodeA,
            nodeB,
            expectedEdge
        );

        thenEdgeIsLoadedAndApproved(expectedEdge);
    }


    @Test
    void itShouldSynchronizeEdgeWhenItAlreadyExistWithoutDuplication() throws GraphException {
        var nodeAId = UniversallyUniqueIdentifier.fromString("10ee31f2-1017-4c71-b4fb-44b52923e6f8");
        var nodeBId = UniversallyUniqueIdentifier.fromString("f7a63c6b-b51d-478d-913f-7ca2bccfcf61");

        var nodeA = new Node(
            nodeAId,
            "Glass"
        );
        var nodeB = new Node(
            nodeBId,
            "Water"
        );
        var synchronizedEdge = new Edge(
            nodeA,
            "contains",
            nodeB
        );

        this.givenSynchronized(
            nodeA,
            nodeB,
            synchronizedEdge
        );

        var sameEdgeWithAttribute = new Edge(
            nodeA,
            "contains",
            nodeB
        );
        sameEdgeWithAttribute = sameEdgeWithAttribute.add(
            new LeafAttribute<>("new", new StringAttributeValue("value")));

        this.whenSynchronize(
            nodeA,
            nodeB,
            sameEdgeWithAttribute
        );
        this.thenEdgeIsLoadedAndApproved(synchronizedEdge);
        this.thenEdgeDoesNotExists(sameEdgeWithAttribute);
    }

    @Test
    void itShouldSynchronizeEdgesWhenSynchronizingNodeByAttribute() throws GraphException {
        var existingNodeId =
            UniversallyUniqueIdentifier.fromString("10ee31f2-1017-4c71-b4fb-44b52923e6f8");

        var existingNode = new Node(
            existingNodeId,
            "node_from_type"
        );
        existingNode = existingNode.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("666"))
        );
        this.givenSynchronized(
            existingNode
        );

        var nodeToBeMerged = new Node("node_from_type");
        nodeToBeMerged = nodeToBeMerged.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("666"
            ))
        );
        this.testNodeIdentificatorsProvider.add(
            nodeToBeMerged.getType(),
            new NodeIdentificator("identifying_attribute")
        );
        var nodeTo = new Node(
            UniversallyUniqueIdentifier.fromString("aceab77c-92c5-4f59-95bc-0210af2edc76"),
            "node_to_type"
        );
        var synchronizedEdge = new Edge(
            UniversallyUniqueIdentifier.fromString("9d4f37cd-6b20-4a9c-b65c-39406c36bf1c"),
            nodeToBeMerged,
            "contains",
            nodeTo
        );
        this.whenSynchronize(
            nodeToBeMerged,
            nodeTo,
            synchronizedEdge
        );
        var expectedEdge = new Edge(
            synchronizedEdge.getId(),
            existingNode,
            "contains",
            nodeTo
        );
        var actualEdge = this.edgeRepository.loadEdge(
            synchronizedEdge.getId(),
            "contains"
        );
        this.thenNodeWithGivenNodeIdDoesNotExist(nodeToBeMerged);
        this.thenEdgesHaveSameIdAndTypeAndNodeIds(
            expectedEdge,
            actualEdge
        );
    }

    @Test
    void itShouldNotMergeDuplicateNodesByIdentificators_WhenTypesDiffer() {
        var nodeA = new Node("typeA");
        this.testNodeIdentificatorsProvider.add(
            nodeA.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        var nodeB = new Node("typeB");
        this.testNodeIdentificatorsProvider.add(
            nodeB.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        nodeA = nodeA.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("same_value")));
        nodeA =
            nodeA.add(new LeafAttribute<>("A_attribute", new StringAttributeValue("A_value")));
        nodeB = nodeB.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("same_value")));
        nodeB =
            nodeB.add(new LeafAttribute<>("B_attribute", new StringAttributeValue("B_value")));

        var graph = new Graph(
            nodeA,
            nodeB
        );

        var actualGraph = whenMergeDuplicates(graph);

        this.thenGraphApproved(
            actualGraph
        );
    }

    @Test
    void itShouldMergeDuplicateNodesByIdentificators() {
        var nodeA = new Node("type");

        var nodeB = new Node("type");

        var nodeC = new Node("type");
        this.testNodeIdentificatorsProvider.add(
            nodeC.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        nodeA = nodeA.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("matched"))
        );
        nodeA = nodeA.add(
            new LeafAttribute<>("A_attribute", new StringAttributeValue("A_value"))
        );
        nodeB = nodeB.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("matched"))
        );
        nodeB = nodeB.add(
            new LeafAttribute<>("B_attribute", new StringAttributeValue("B_value"))
        );
        nodeC = nodeC.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("not_matched"))
        );

        var graph = new Graph(
            nodeA,
            nodeC,
            nodeB
        );

        var actualGraph = whenMergeDuplicates(graph);

        thenGraphApproved(actualGraph);
    }

    @Test
    void itShouldMergeDuplicateNodesByIdentificators_AndItWillFixEdges() {
        var nodeA = new Node("type");
        this.testNodeIdentificatorsProvider.add(
            nodeA.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        var nodeB = new Node("type");
        this.testNodeIdentificatorsProvider.add(
            nodeB.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        var nodeC = new Node("type_C");
        var edge = new Edge(
            nodeB,
            "connects",
            nodeC
        );

        nodeA = nodeA.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("matched")));
        nodeA =
            nodeA.add(new LeafAttribute<>("A_attribute", new StringAttributeValue("A_value")));
        nodeB = nodeB.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("matched")));
        nodeB =
            nodeB.add(new LeafAttribute<>("B_attribute", new StringAttributeValue("B_value")));

        var graph = new Graph(
            nodeA,
            nodeB,
            nodeC,
            edge
        );

        var actualGraph = whenMergeDuplicates(graph);

        this.thenGraphApproved(actualGraph);
    }

    @Test
    void itShouldMergeDuplicateNodesByIdentificators_AndItWillMergeEdges() {
        var nodeA = new Node("type");
        this.testNodeIdentificatorsProvider.add(
            nodeA.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        var nodeB = new Node("type");
        this.testNodeIdentificatorsProvider.add(
            nodeB.getType(),
            new NodeIdentificator("identifying_attribute")
        );

        var nodeC = new Node("type_C");
        var e1 = new Edge(
            nodeA,
            "connects",
            nodeC
        );
        e1 = e1.add(
            new LeafAttribute<>("A_edge_attribute", new StringAttributeValue("A_edge_value")));
        var e2 = new Edge(
            nodeB,
            "connects",
            nodeC
        );
        e2 = e2.add(
            new LeafAttribute<>("B_edge_attribute", new StringAttributeValue("B_edge_value")));

        nodeA = nodeA.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("matched")));
        nodeA =
            nodeA.add(new LeafAttribute<>("A_attribute", new StringAttributeValue("A_value")));
        nodeB = nodeB.add(
            new LeafAttribute<>("identifying_attribute", new StringAttributeValue("matched")));
        nodeB =
            nodeB.add(new LeafAttribute<>("B_attribute", new StringAttributeValue("B_value")));

        var graph = new Graph(
            nodeA,
            nodeB,
            nodeC,
            e1,
            e2
        );

        var actualGraph = whenMergeDuplicates(graph);

        this.thenGraphApproved(actualGraph);
    }
    
    @NotNull
    private Graph whenMergeDuplicates(Graph graph) {
        var actualGraph = graph.traversable();
        return this.identifyingGraphSynchronizer
            .mergeDuplicateNodesByIdentificators(actualGraph)
            .getGraph();
    }

    private void givenSynchronized(AttributeContainer... attributeContainers) {
        this.synchronize(attributeContainers);
    }

    private void whenSynchronize(AttributeContainer... attributeContainers) {
        this.synchronize(attributeContainers);
    }

    private void thenEdgeIsLoadedAndApproved(Edge expectedEdge) {
        var actualEdge = this.edgeRepository.loadEdge(
            expectedEdge.getId(),
            expectedEdge.getType()
        );

        this.thenEdgeApproved(actualEdge);
    }

    private void thenEdgeDoesNotExists(Edge edge) {
        var id = edge.getId();
        var type = edge.getType();
        assertThrows(
            EdgeNotFound.class,
            () -> this.edgeRepository.loadEdge(id, type)
        );
    }

    private void synchronize(AttributeContainer[] attributeContainers) {
        var graph = new Graph(attributeContainers);
        this.identifyingGraphSynchronizer.synchronize(graph);
    }

    protected void thenNodeWithGivenNodeIdDoesNotExist(Node nodeToBeMerged) {
        var exists = this.nodeRepository.nodeExists(
            nodeToBeMerged.getId(),
            nodeToBeMerged.getType()
        );
        Assertions.assertFalse(
            exists,
            String.format(
                "Node with id %s and type \"%s\" was expected to not exist, but it actuallu exists.",
                nodeToBeMerged.getId(),
                nodeToBeMerged.getType()
            )
        );
    }
}
