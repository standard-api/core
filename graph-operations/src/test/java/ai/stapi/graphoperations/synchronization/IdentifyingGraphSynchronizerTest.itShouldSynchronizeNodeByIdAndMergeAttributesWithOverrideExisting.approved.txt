node_type: Test_node_type
node_hash: 7E640
node_edges:
attributes:
	string_type -> test value
	string_type -> test value_overwritten
	string_type2 -> test value2
